// @aws-amplify/graphql-conversation-transformer/lib/resolvers/verify-session-owner-resolver.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.verifySessionOwnerMappingTemplate = void 0;
const graphql_transformer_core_1 = require("@aws-amplify/graphql-transformer-core");
const ts_dedent_1 = require("ts-dedent");
const verifySessionOwnerMappingTemplate = () => {
    const req = createVerifySessionOwnerRequestFunction();
    const res = createVerifySessionOwnerResponseFunction();
    return { req, res };
};
exports.verifySessionOwnerMappingTemplate = verifySessionOwnerMappingTemplate;
const createVerifySessionOwnerRequestFunction = () => {
    const requestFunctionString = `
      export function request(ctx) {
        const { authFilter } = ctx.stash;

        const query = {
          expression: 'id = :id',
          expressionValues: util.dynamodb.toMapValues({
            ':id': ctx.args.conversationId
          })
        };

        const filter = JSON.parse(util.transform.toDynamoDBFilterExpression(authFilter));

        return {
          operation: 'Query',
          query,
          filter
        };
      }
      `;
    return graphql_transformer_core_1.MappingTemplate.inlineTemplateFromString((0, ts_dedent_1.dedent)(requestFunctionString));
};
const createVerifySessionOwnerResponseFunction = () => {
    const responseFunctionString = `
      export function response(ctx) {
        if (ctx.error) {
          util.error(ctx.error.message, ctx.error.type);
        }

        if (ctx.result.items.length !== 0) {
          return ctx.result.items[0];
        }

        util.error('Conversation not found', 'ResourceNotFound');
      }`;
    return graphql_transformer_core_1.MappingTemplate.inlineTemplateFromString((0, ts_dedent_1.dedent)(responseFunctionString));
};
